### ISSUE TITLE
Performance compared to gdx-box2d

### ISSUE THREAD

**Issue by soylomass** [dyn4j/dyn4j#99]
I tried this library almost 2 years ago to replace gdx-box2d which uses the original c++ library with JNI, and had to give up because the performance was much slower in my application (a game server with 6000+ bodies, most of which are asleep most of the time), where physics step are called 20 times a second therefore I need a world step time of around 10ms.

Has the performance improved in this time? Is it comparable to the original c++ library?

Thanks in advance.

**Comment by mtsamis** [comment]
While I have not compared the speed to gdx-box2d specifically, there have been much performance optimizations to dyn4j in the last year or so. You should definitely try it again now (be sure to use 3.4.0 or the latest source).

Since you're talking about 6000+ bodies, [was a bug previously with large worlds that now is fixed](https://github.com/dyn4j/dyn4j/issues/39).
I also suggest you use the new (though somewhat experimental at this point) [broadphase detector](https://github.com/dyn4j/dyn4j/pull/49) that is included in dyn4j which should be much faster especially for big worlds.
There have been various other optimizations as well, see the [release notes](https://github.com/dyn4j/dyn4j/blob/master/release-notes.txt)

Do you use CCD? With CCD turned off + the suggestions above I think you can achieve much better performance than 10ms/step.

Cheers

**Comment by soylomass** [comment]
That's good to hear, I can't wait to use a pure Java 2D physics engine. JNI Box2D works great but whenever something fails it's almost impossible to debug and native exceptions make the whole app crash.

About CCD, I don't use it, but some fast moving objects objects are temporarily set as bullet to improve their collision detection.

I'll give it a try again soon.

**Comment by soylomass** [comment]
@mtsamis 

`@Override
	public boolean contains(E collidable) {
		int size = collidable.getFixtureCount();
		
		for (int i = 0; i < size; i++) {
			T fixture = collidable.getFixture(i);
			
			if (!this.contains(collidable, fixture)) {
				return false;
			}
		}
		
		return true;
	}`

I was taking a look at the new Broadphase code. This piece returns false if the first fixture wasn't found. Shouldn't it return false only if all of them weren't found?

**Comment by mtsamis** [comment]
Hey, I think this is the intended behavior as per the javadoc
```
* Returns true if all the {@link Fixture}s on the given {@link Collidable}
* have been added to this broad-phase.
```
So if even one fixture is not added (say the first) then this method must return false.

**Comment by soylomass** [comment]
I've been doing some tests in another game, here are the results:
Using 3.4.0, the performance did improve greatly.

The game is raaaaft.io. I created a local world without characters, only AI. If you open the game, there is a big world, with islands and AI animals. Around 700 bodies.

Results: https://imgur.com/a/FRb9D19

As you can see in the results, 3.4.0 did bring the performance closer to Box2D, but the latter is still around 50% faster.
Also, the results with Box2D are more stable, the step time is almost always the same, while with Dyn4J, sometimes it's well below 1ms and sometimes close to 4ms.

I also get better results with DynamicAABBTree.

Do you have any idea looking at the JProfiler screenshots on what could be done to improve the performance even further?

**Comment by wnbittle** [comment]
This is great feedback - love the profiler screenshots.

Does your game need Continuous Collision Detection (CCD)?  This is used for fast moving bodies and tunneling.  If not, I would turn it off globally.

Apart from that, there’s probably some work that could be done to get CCD faster.  Nothing else seemed to jump out at me right away.

Another 50% could be challenging for sure, but these profiling snapshots from a real application are vital to continuously improve.

**Comment by mtsamis** [comment]
> About CCD, I don't use it, but some fast moving objects objects are temporarily set as bullet to improve their collision detection.

OP uses CCD conditionally for some bullet objects.

As William said there are improvements for CCD to be done that can give you that extra performance you want. I actually have some code that greatly speeds up CCD, it just hasn’t made it into dyn4j yet because we haven’t found some good solution for https://github.com/dyn4j/dyn4j/issues/63. If you really want it I can write the code in some separate branch for you to use.

Also can you give some insight about the specific setup of your objects? How many bodies are static? Are the bodies spread out of concentrated? What shapes you use? What about joints? Etc 

There may be more specific things to do given more data.
Also to make comparison fair, I think jbox2d uses float precision (at least box2D does so) whereas dyn4j uses double precision everywhere. So you get some extra precision and stability. There is a measurable cpu and memory cost for using double precision so I would say 30% slower is quite comparable to jbox2d at the moment!

Finally about the cpu spikes, maybe those are caused by the occasional extra work the JVM has to do? (ie optimizations, garbage etc)
Is there something else that happens on those frames that may be the culprit?

**Comment by soylomass** [comment]
EDIT: I noticed that ContinuousDetectionMode was set to ALL by default, is that normal? After I set it to BULLET_ONLY, the spikes were no more and the step time is on level with Box2D

Note that it's not jBox2d, it's GDX-Box2D, which uses the C++ library though JNI. I've tested Java ports in the past but none could achieve the same performance.

Here is an analysis of the status of the world at a random moment:
#### ENT. ANALYISIS: 550 can sleep, 86 sleeping, 0 inactive, 0 bullet, 309 static, 0 kinematic, 352 dynamic #####

Most objects are awake but that's ok, because other than Islands and Trees, the rest are all animals, which mostly don't sleep, and animals walking above Islands keep Islands awake too.

I will do more testing, but it looks promising! You did a great job with the library

**Comment by wnbittle** [comment]
Wow, that's great to hear!  I _think_ I did this to improve adoption for new users, so they don't need to worry about tunneling.  More advanced users could then back it down to BULLET_ONLY and tag their specific bodies that need it.  @mtsamis might have an opinion here as well.  I'm open to changing the default though.

With respect to the library and performance - thanks!  It's not often I hear good feedback.  In the past year, @mtsamis has contributed significantly to the performance and has been a good driving force behind general improvements.

**Comment by mtsamis** [comment]
Thanks for you kind words and all your contributions as well, William!

In the past I also believed that it would be good to change the default CCD mode to something else, but then I found out that the CCD phase can be greatly optimized and then it won’t be a problem. Unfortunately I haven’t been able to make a clean implementation of that yet, but we can add it to the roadmap. 

@soylomass I guess you can start migrating to dyn4j now that the performance is as good! :)

**Comment by soylomass** [comment]
@mtsamis I will! The difference between my main game and Raaaaft, is that unlike the latter, where I built the physics using interfaces to make it easy to exchange physics engines, in my main game Box2D is hardcoded everywhere.
I want to do it just like in Raaaaft, that way I can switch between Dyn4j and Box2D easily to compare performance. It will take me some time as it's a big project, but when it's done I'll come back with the results.

Thanks!

**Comment by soylomass** [comment]
Good and bad news.
The good news is I finished making my game "multi-engine", so now I can switch back and forth from Box2D to Dyn4j to test performance.

The bad news is that in this game, the performance with Dyn4j got greatly reduced:

https://imgur.com/a/L8A9isF

The game in this case is https://deeeep.io

I created a local world with everything but Animals. So there is the terrain, the props, and food (most of those are Static, except for some moving foods)

In one of the tests in the link I removed most moving foods to make the number of Dynamic bodies close to 0.

As you can see, even with almost every body being Static, the performance is much slower than Box2D.

Is it related to language differences or is it related to implementation?

Do you have any idea on why in the other game the performance was similar to Box2d and in this case there is this big of a difference?

**Comment by mtsamis** [comment]
Hey! I think can provide some answers as I have been working on a box2D fork lately.

> Is it related to language differences or is it related to implementation?
This is related to the implementation not the language, hence we can probably fix it :)

Box2D is quite efficient at handling static bodies and there’s a key difference here: box2D can persist collisions/contacts of static bodies without testing them in the broadphase at all.
Dyn4j will perform broadphase collision for all bodies whereas box2D only for the dynamic ones.

Since you simulation consists mostly of static bodies you get a huge performance penalty with dyn4j. 

Also another thing to keep in mind (I think I haven’t mentioned this before): LazyAABBTree was designed/optimized for the inverse scenario, worlds with mostly dynamic bodies and a lot of collisions.

@wnbittle although the contact manager is different from box2D it should be possible to use the same strategy, right? Unfortunately I don't have a computer where I can code right now, but I can help with the design and provide references to the corresponding box2D code.

P.S. if anyone is working/experimenting with Box2D I’m working on a fork that may be of interest.

**Comment by wnbittle** [comment]
@soylomass do you need the output from the static vs. static collision tests?  If no, the next question is to make sure the term "static" means what i think you mean.  There's technically 3 types of objects: Static, Kinematic, and Dynamic.  Static/Kinematic are treated the same in dyn4j.

That said, if you need to control what gets collision detected you can use the BroadphaseFilter interface to fine tune this.  The DefaultBroadphaseFilter is only looking at the Fixture Filters.  Would this be an option?

@mtsamis can you provide more details about the caching mechanism for the broadphase in box2d?

**Comment by soylomass** [comment]
@wnbittle In Box2D static-static and static-kinematic don't collide and don't trigger collision events, so I got used to that.

I'll take a look at BroadphaseFilter.

**Comment by soylomass** [comment]
The low performance happens even if I return false to every method in the BroadphaseFilter, so it cannot be improved from there.

**Comment by wnbittle** [comment]
@soylomass thanks.  Can you provide some more profiling screenshots?  From the last ones you posted it was clearly CCD that was the problem - they were very helpful and tracking down the exact place where performance was bad.

**Comment by mtsamis** [comment]
Box2d queries only moving objects in the tree (hence doesn’t query static bodies at all):
https://github.com/erincatto/box2d/blob/60d36c498bc86afa0cd140d7b9fc37d5df76a39a/include/box2d/b2_broad_phase.h#L178

Contacts are maintained from the previous frame and are removed if their AABBs cease to overlap:
https://github.com/erincatto/box2d/blob/60d36c498bc86afa0cd140d7b9fc37d5df76a39a/src/dynamics/b2_contact_manager.cpp#L161

Since only moving objects are tested in the broadphase it’s logical that in OP’s case where there are only/mostly non-moving objects performance is superior. There are some more ways to achieve this contact caching.

**Comment by mtsamis** [comment]
> Can you provide some more profiling screenshots? From the last ones you posted it was clearly CCD that was the problem - they were very helpful and tracking down the exact place where performance was bad.

> https://imgur.com/a/L8A9isF

@wnbittle I think it’s clear the broadphase detect phase is the problem? Everything else is a tiny fraction in the profiler, toi, narrow phase and contact manager are all there but take up a tiny fraction. Is something missing versus the previous screenshots that I miss?

**Comment by soylomass** [comment]
@wnbittle I posted new screenshots in the last report: https://imgur.com/a/L8A9isF
Those are ok or you mean new screenshots with the BroadphaseFilter returning false to all?

**Comment by wnbittle** [comment]
@soylomass @mtsamis sorry guys - I missed the link to the new profiling screenshots so I was behind on the conversation - doh!  Yeah, it's clearly the Broadphase and it's clear that what I suggested won't help you.

The broadphase is a tree - I'd be interested to know what the height of the tree is.  I don't have a public method for this unfortunately.  Another useful piece of data would be the number of iterations the detectNonRecursive method performs.  Just want to confirm that the tree itself isn't the problem.  I'll try to replicate this in a unit test as well.

@mtsamis I see that now.  So he's caching all the broadphase collision pairs and tracking what was updated (moved) and then only rechecking those that moved.  I think the current BroadphaseDetector interface would support that type of tracking in our existing broadphase implementations without any breaking changes.  The update method indicates that something has changed, we'd need to only compare whether the AABB is different than the last, mark that as a body to retest in the next run, then do detection on those only.

**Comment by mtsamis** [comment]
@dyn4j Yeah I think with some minor refactoring this useful optimization can make it into dyn4j as well. We must see what’s the best way to implement this based on dyn4j design and goals. 

@soylomass it would be nice if you could share something to reproduce the scene, so we can perform more in depth investigation.

**Comment by soylomass** [comment]
@mtsamis I'll do soon.

**Comment by wnbittle** [comment]
@soylomass @mtsamis Just a quick update.  

I've played around with some implementations of this optimization and it's around 33% faster in my test scenario (evenly distributed 6k static bodies with random shapes and sizes and 100 dynamic circle bodies).  The challenge is implementing the cache without negating the benefit of the optimization _within the current design of the library_.

box2d has a contact caching mechanism at the engine level - i.e. it's shared by the collision and dynamics pipelines.  In addition to this, the box2d broadphase is only designed to output **NEW** collision pairs on each update.  It goes something like this:

1. Perform Broadphase to detect NEW pairs by only checking those that moved (the optimization we're interested in)
2. Add new pairs to contact manager (duplicate detection happens here)
3. For ALL pairs check AABB overlap (needed since the broadphase doesn't remove pairs)
4. For remaining pairs do narrowphase + manifold generation

dyn4j on the other hand only has caching at the contact constraint level.  On top of this, the dyn4j broadphases are designed to output all overlaps every frame.  It goes something like this:

1. Perform Broadphase to detect ALL pairs
2. For all pairs do Narrowphase
3. For all narrowphase results, do manifold generation
4. For all manifolds, add to contact manager (caching is here)

In the end, the process is the same, apart from the differences in the broadphase.

@soylomass What's 33% increase going to do for you?  Would that be enough to switch - I get the feeling the answer is no.

**Comment by soylomass** [comment]
@wnbittle I would need at least the same performance as with Box2D to switch honestly (**for this game**) as the screenshots I showed are from the server running on my computer, however the VPS where the gameservers are hosted have slower CPUs and I'm already running at the limit of it.

**Comment by wnbittle** [comment]
@soylomass  totally understand.  The 30% increase was from a naive solution just to understand what the potential impact of that particular optimization would be - I think we can do better.  However, to do better I think we'll need to make some fundamental design changes to the library.  I'm not opposed to any of that - in fact I've been thinking of some for a while now.

As you might imagine, this wouldn't be overnight - and that's very disappointing to me - as this affects you and other would-be adopters.  On the other hand, I'm challenged to make progress on this front and hopefully through further commitment we can obtain this goal.

**Comment by mtsamis** [comment]
+1 from me, we will make more optimizations and it is evident that there is potential to greatly optimize further for static bodies, but this will need time to design and implement. 
At the moment I’m taking a deep dive into broadphase collision and there are a lot of improvements to do there overall. 

@soylomass since the bottleneck is very specifically in the broadphase you can always implement your own detector that is efficient for your exact case and is greatly optimized towards static bodies. Otherwise make sure to keep an eye here until we improve the general one sufficiently :)

**Comment by wnbittle** [comment]
@mtsamis @soylomass It's been a while and I've been busy.  I've added this optimization and a number of others.  In my simulations (6000 static bodies, 100 dynamic bodies) I'm seeing anywhere from 30-40% performance improvement.

I've also redesigned a good bit of the library.  Thankfully almost every change will be marked with deprecation comments to know how to upgrade and a lot of it won't affect typical users of the library.  The basic goals were to:

1. Add this optimization and a few others I've been thinking about
2. Further segment the code to improve maintainability

I still have a good bit of work left to do, but I'm happy with the result so far.  I don't have anything for you to test yet, but will be there soon.

@mtsamis I'm not sure what to do with the LazyAABBTree broadphase.  The optimization depends on 1) AABB expansion to know something has updated and 2) only running a portion of the broadphase (updated only).  Does it still make sense to keep around?  For now I've just updated to work, but it's not taking advantage of the improvements.

**Comment by mtsamis** [comment]
Hey William, it’s great to hear you further optimized and improved code design in dyn4j!

About the LazyAABBTree:
I have made some much better attempts for broadphase algorithms since then so it’s kind of outdated; I’m currently experimenting with broadphase algorithms in Box2D and I was going to propose a new  implementation to replace the LazyAABBTree.
For now, based on what you say I can imagine it still being faster for highly dynamic scenes composed mostly of colliding dynamic bodies but not much else. 

I see you added the 4.0.0 milestone and as you say there is more work to do, so: LazyAABBTree can probably be removed(/replaced) for 4.0.0. After I see your new changes and finish my experiments with Box2D broadphases I plan to propose a new implementation for that.

It would be cool if you have all the new stuff somewhere to browse, so I can keep track of the progress (and help if possible)! If the code is still under work and not fit for the master branch how about making a new 4.0.0 branch?

**Comment by soylomass** [comment]
That's good to hear! I'm ready to benchmark whenever improvements are ready.

**Comment by wnbittle** [comment]
@mtsamis I've checked in all the changes so far into a new branch called "v4".  Most of the code is complete, but there's a good bit of comments left to create/update and all the JUnit tests need to be updated from any deprecated methods.

**Comment by assofohdz** [comment]
Any specific params I need to know about if I want to test this v4 ?

**Comment by wnbittle** [comment]
Nothing specific just follow the deprecation comments to use the new classes.  Most things will be named the same.  You might see more generics and few API changes as well. There's been a few bug fixes and a few behavioral changes, but nothing crazy.  Just keep in mind that the branch isn't well tested yet.

**Comment by assofohdz** [comment]
> Nothing specific just follow the deprecation comments to use the new classes. Most things will be named the same. You might see more generics and few API changes as well. There's been a few bug fixes and a few behavioral changes, but nothing crazy. Just keep in mind that the branch isn't well tested yet.

Im missing the world.addListener - is there a replacement?

**Comment by assofohdz** [comment]
Specifically adding CollisionLIstener :) Looking forward to seeing how this runs

**Comment by wnbittle** [comment]
Yeah, you need to add them via getCollisionListeners().add(...) now.

**Comment by assofohdz** [comment]
> Yeah, you need to add them via getCollisionListeners().add(...) now.

Thank you :)

**Comment by assofohdz** [comment]
Nothing broke 👍  - will report back with findings as they come

**Comment by assofohdz** [comment]
I opted for a built-in phys-system. Too much overhead in marshalling between dyn4j and the engine. I do like dyn4j very much though. Best of luck testing these optimizations.

**Comment by wnbittle** [comment]
@assofohdz Thanks for testing anyway.  Care to elaborate on your decision?  I'd be interested to know what the replacement was (I'm guessing "the engine" is what has a built-in physics system) and the specifics of the marshalling overhead.

**Comment by assofohdz** [comment]
Lots of the systems in the engine are already hooked up on an abstract form of a "body" - and even though it was meant to be easily replaceable (replacing Collision Handlers etc. with something external) it just quickly became a spiderweb to unwrangle in order to use Dyn4Js World and Body. If it does become easier to replace along the way, I'm keen to go back to Dyn4J, but as it is - the built-ins already take care of unloading/loading sleepy bodies if they are not nearby any active objects - making it very optimized.  It doesn't have any definition of broadphase/narrowphase, but the game i'm building is super simple in terms of physics, so in all likelyhood I do not need this kind of collision testing. 

When I mention marshalling its because the engine uses Vector3 doubles in the backend and Dyn4J (rightly so) uses Vector2 - so everywhere I integrate the two systems, I have to translate between the two data structures

I was actually hoping that the built-ins would allow me to easily choose my own physics lib, in which case I would easily have gone with Dyn4J (given that we could get these optimizations to work)

**Comment by wnbittle** [comment]
@assofohdz Thanks for the additional detail - any feedback to improve the quality/utility of the library is important to me.  What's the name of the engine you are working with (or is it in house)?

**Comment by assofohdz** [comment]
The engine is jMonkeyEngine - but the built-ins I mentioned is part of a private framework. jME doesn't have a native physics library per say, but refer to Dyn4J/bullet when needed.

I did use Dyn4J up until I switched to this private framework, so it can absolutely work - this framework just provides a ton of other entity management stuff I want to work with.

**Comment by assofohdz** [comment]
@wnbittle how is the performance boosts coming along ? I may run into issues with my current physics engine and may divert back to Dyn4J

**Comment by wnbittle** [comment]
@assofohdz All the performance enhancements are complete.  What remains is testing, which I'm making good progress on.  I've added ~200 unit tests and fixed a number of bugs this way.  Some API is still stabilizing.

**Comment by wnbittle** [comment]
Another update.  Still working through API refinement, bugs, and todos.

Found another interesting detail in Box2d - the broadphase contains the _swept_ AABBs, not just the expanded AABBs.  I've thought about this in the past as a way to improve CCD performance, but not in this way - I was thinking about it being a separate instance of the broadphase that would need to be maintained.

This would be great for most worlds where there's a few fast moving objects.  There are some details though.  If everything is moving fast, then we basically have what we have today with CCD, but slower because there's more narrowphase collision detection.  Fast moving objects will have huge AABBs which means more narrowphase collision detection, but if the number of these is small, then it shouldn't matter too much and the savings for CCD could be huge.  As an object's speed slows, their AABB will need to be reduced so that you don't keep huge AABBs out there overlapping everything.

I think I'm going to leave the swept AABB enhancement for a later release.  There shouldn't be too many breaking changes to the broadphase to support it and even those breaking changes wouldn't break those not using them directly.  What would change significantly is the CCD process - we could then use the broadphase results to know for sure what pairs to test.

**Comment by mtsamis** [comment]
William, One comment from me for the swept AABBs in Box2D, because I have measured the impacts of this specific design choice. The swept AABBs make simulation time more unstable and cause even bigger performance spikes: They essentially optimize the 'easy' parts of a simulation (slow moving bodies, few collisions, sparse simulations) and come with a huge performance penalty on the 'difficult' (a lot of collisions, dense simulations, ...). Since dyn4j aims to be used real-time mostly this is a bad thing. It's in a sense a better tradeoff to have higher average time and lower max running time than low running time and huge spikes.

If you want to study this in Box2D a perfect example is the 'Add Pair' demo (CCD off). This simulation is easy on the CPU except from some ~5 frames. you can see that the swept aabbs create tons of false-positive pairs and kill performance. If you slightly increase the body count there the result is even more obvious and the simulations takes seconds per frame to run (for no obvious reason).

All the above apply for when global CCD is off. For when CCD is on doing this should bring great performance benefits by avoiding the current double loop to find CCD pairs and getting them directly from the broadphase (though this can be done seperately anyway, as I had suggested in the past).

Tl;dr Box2D uses the swept AABBs even when CCD is off, which will be a bad performance hit here. But when CCD is on this will bring big benefits, especially for dyn4j. So the obvious suggestion is to somehow enable swept AABBs conditionally?

**Comment by wnbittle** [comment]
@mtsamis thanks for the commentary - all good to know.  I've added this as an enhancement to evaluate (among others).

@soylomass Getting close to deployment of 4.0.0.  Hopefully you're still interested in doing some evaluation when that happens.

**Comment by soylomass** [comment]
@wnbittle I am! Looking forward to it.

**Comment by wnbittle** [comment]
Closing this for v4 deployment.  Once deployed and tested, feel free to open another issue if performance is still a concern.

**Comment by wnbittle** [comment]
Version 4.0.0 has been released.
